### PR TITLE
Implement static call

### PR DESCRIPTION
**Pull request by norhh**

### CODE REVIEW

File: mythril/laser/ethereum/instructions.py [hunk 1]
```diff
@@ -68,14 +69,15 @@ class StateTransition(object):
     if `increment_pc=True`.
     """
 
-    def __init__(self, increment_pc=True, enable_gas=True):
+    def __init__(self, increment_pc=True, enable_gas=True, static_check=False):
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instructions.py]: Can you change the variable name to something that is more clear about the purpose.

File: mythril/laser/ethereum/instructions.py [hunk 2]
```diff
@@ -1854,6 +1864,21 @@ def call_(self, global_state: GlobalState) -> List[GlobalState]:
             )
             return [global_state]
 
+        if environment.static:
+            if isinstance(value, int) and value > 0:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instructions.py]: I don't think this is possible at a transaction creation level

File: mythril/laser/ethereum/state/environment.py [hunk 3]
```diff
@@ -49,6 +50,7 @@ def __init__(
         self.gasprice = gasprice
         self.origin = origin
         self.callvalue = callvalue
+        self.static = static
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/environment.py]: Variable name not clear.
How about `write_permission`

File: mythril/laser/ethereum/instructions.py (hunk 2 above)
Comment (norhh) [reply mythril/laser/ethereum/instructions.py]: meaning?

File: mythril/laser/ethereum/state/environment.py (hunk 3 above)
Comment (norhh) [reply mythril/laser/ethereum/state/environment.py]: Variable name is good right? Like it says something like `environment is static` if static is true,

File: mythril/laser/ethereum/instructions.py [hunk 4]
```diff
@@ -68,14 +69,18 @@ class StateTransition(object):
...
+    ):
         """
 
         :param increment_pc:
         :param enable_gas:
+        :param state_mutation_func: The function mutates state
         """
         self.increment_pc = increment_pc
         self.enable_gas = enable_gas
+        self.state_mutation_instruction = state_mutation_instruction
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instructions.py]:
```suggestion
        self.is_state_mutation_instruction = is_state_mutation_instruction
```

File: mythril/laser/ethereum/instructions.py (hunk 2 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/instructions.py]: Ah nvm, github code folding had me confused

File: mythril/laser/ethereum/instructions.py [hunk 5]
```diff
@@ -1871,20 +1899,70 @@ def call_(self, global_state: GlobalState) -> List[GlobalState]:
...
     def call_post(self, global_state: GlobalState) -> List[GlobalState]:
         """
 
+        :param global_state:
+        :return:
+        """
+
+        return self.post_handler(global_state, function_name="call")
+
+    @StateTransition()
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instructions.py]: This is also a state mutating instruction according to the yellowpaper

Comment (norhh) [reply mythril/laser/ethereum/instructions.py]: But the EIP calls it some exception https://github.com/ethereum/EIPs/blob/master/EIPS/eip-214.md .

File: mythril/laser/ethereum/state/environment.py (hunk 3 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/state/environment.py]: Its just that in the yellow paper this variable is called $$I_w$$ and the description is "the permission to make modifications to the state". 
The way to get this value to be true is "STATICCALL", which makes the name `static` a reasonable one. 
However, I feel like write_permission or has_write_permission directly confers what it is that the variable describes. 
Anyway, this is not really a dealbreaker 😅

Comment (JoranHonig) [reply mythril/laser/ethereum/state/environment.py]: Lol nvm, documentation as always is not consistent. In other places they call the flag a static flag, so we can reasonably assume that anybody should be able to understand the variable name

File: mythril/laser/ethereum/instructions.py (hunk 5 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/instructions.py]: I've checked the geth and parity implementations an the EIP is indeed what is implemented. 
So the status quo is good 👍

**Comment by palkeo** [comment]
Is there anything preventing this PR to be merged?

(bug #968 )

File: mythril/laser/ethereum/instructions.py [hunk 6]
```diff
@@ -1894,17 +1971,16 @@ def callcode_post(self, global_state: GlobalState) -> List[GlobalState]:
...
 
         :param global_state:
         :return:
         """
         instr = global_state.get_current_instruction()
         environment = global_state.environment
-
         try:
-            callee_address, callee_account, call_data, value, gas, _, _ = get_call_parameters(
+            callee_address, callee_account, call_data, value, gas, memory_out_offset, memory_out_size = get_call_parameters(
```
Comment (nbanmp) [inline mythril/laser/ethereum/instructions.py]: The EIP https://github.com/ethereum/EIPs/pull/214/files#diff-29d8cb8d19a769c141160ce5c1e65caeR31 claims that there are only 6 parameters for STATICCALL and value is always 0.

Comment (norhh) [reply mythril/laser/ethereum/instructions.py]: In this case it returns 0. there is the parameter `with_value` which is False in this case.

**Pull request merged**